packages/compass-import-export/src/import/import-csv.ts

```diff
@@ -0,0 +1,122 @@
+import type { DataService } from 'mongodb-data-service';
+
+import { createCollectionWriteStream } from '../utils/collection-stream';
+import type { CollectionStreamStats } from '../utils/collection-stream';
+import type { Delimiter, IncludedFields, PathPart } from '../utils/csv';
+import { makeDoc, parseHeaderName } from '../utils/csv';
+import { createDebug } from '../utils/logger';
+
+const debug = createDebug('import-csv');
+
+type ImportCSVOptions = {
+  dataService: DataService;
+  ns: string;
+  input: Readable;
+  abortSignal?: AbortSignal;
+  progressCallback?: (index: number) => void;
+  delimiter?: Delimiter;
+  ignoreEmptyStrings?: boolean;
+  stopOnErrors?: boolean;
+  fields: IncludedFields; // the type chosen by the user to make each field
+};
+
+type ImportCSVResult = CollectionStreamStats & { aborted?: boolean };
+
+export function importCSV({
+  dataService,
+  ns,
+  input,
+  abortSignal,
+  progressCallback,
+  delimiter = ',',
+  ignoreEmptyStrings,
+  stopOnErrors,
+  fields,
+}: ImportCSVOptions): Promise<ImportCSVResult> {
+  debug('importCSV()', { ns: toNS(ns) });
+
+  let numProcessed = 0;
+  const headerFields: string[] = []; // will be filled via transformHeader callback below
+  let parsedHeader: Record<string, PathPart[]>;
+
+  const docStream = new Transform({
+    objectMode: true,
+    transform: function (chunk: Record<string, string>, encoding, callback) {
+      ++numProcessed;
+
+      debug('importCSV:transform', numProcessed, {
+        headerFields,
+        chunk,
+        fields,
+        ignoreEmptyStrings,
+        encoding,
+      });
+
+      if (abortSignal?.aborted) {
+        docStream.destroy();
+      }
+
+      if (!parsedHeader) {
+        parsedHeader = {};
+        for (const name of headerFields) {
+          parsedHeader[name] = parseHeaderName(name);
+        }
+      }
+
+      try {
+        const doc = makeDoc(chunk, headerFields, parsedHeader, fields, {
+          ignoreEmptyStrings,
+        });
+        debug('transform', doc);
+        progressCallback && progressCallback(numProcessed);
+        callback(null, doc);
+      } catch (err: unknown) {
+        // rethrow with the row index appended to aid debugging
+        (err as Error).message = `${
+          (err as Error).message
+        }[Row ${numProcessed}]`;
+        debug('transform error', (err as Error).message);
+        callback(err as Error);
+      }
+    },
+  });
+
+  const collectionStream = createCollectionWriteStream(
+    dataService,
+    ns,
+    stopOnErrors ?? false
+  );
+
+  const parseStream = Papa.parse(Papa.NODE_STREAM_INPUT, {
+    delimiter,
+    header: true,
+    transformHeader: function (header: string, index: number): string {
+      debug('importCSV:transformHeader', header, index);
+      headerFields.push(header);
+      return header;
+    },
+  });
+
+  return new Promise((resolve, reject) => {
+    pipeline(input, parseStream, docStream, collectionStream, function (err) {
```

Comment:
Apparently pipeline can return a promise and take the abort signal in a last-param options hash, but typescript is NOT happy if I try and do that. My best guess is that the type of collectionStream is confusing something.

Reply:
I think you’d want the promise version of `pipeline` from `import('stream').pipeline`, not the one from `import('stream')`: https://nodejs.org/api/stream.html#streams-promises-api (i.e. not https://nodejs.org/api/stream.html#streampipelinesource-transforms-destination-callback)

Reply:
Hmm still the same error. I'm just gonna go with what I have for now as it works.

Reply:
Btw, this compiles fine for me (it’s just a bit weird/annoying that it doesn’t consider the `signal` option to be optional in the options argument):

```diff
diff --git a/packages/compass-import-export/src/import/import-csv.ts b/packages/compass-import-export/src/import/import-csv.ts
index 63c5fc89cb65..f35d006a1c1a 100644
--- a/packages/compass-import-export/src/import/import-csv.ts
+++ b/packages/compass-import-export/src/import/import-csv.ts
@@ -1,4 +1,4 @@
-import { Transform, pipeline } from 'stream';
+import { Transform, promises as streamPromises } from 'stream';
 import type { Readable } from 'stream';
 import Papa from 'papaparse';
 import toNS from 'mongodb-ns';
@@ -10,6 +10,7 @@ import type { Delimiter, IncludedFields, PathPart } from '../utils/csv';
 import { makeDoc, parseHeaderName } from '../utils/csv';
 import { createDebug } from '../utils/logger';
 
+const { pipeline } = streamPromises;
 const debug = createDebug('import-csv');
 
 type ImportCSVOptions = {
@@ -26,7 +27,7 @@ type ImportCSVOptions = {
 
 type ImportCSVResult = CollectionStreamStats & { aborted?: boolean };
 
-export function importCSV({
+export async function importCSV({
   dataService,
   ns,
   input,
@@ -101,22 +102,17 @@ export function importCSV({
     },
   });
 
-  return new Promise((resolve, reject) => {
-    pipeline(input, parseStream, docStream, collectionStream, function (err) {
-      if (err) {
-        if (err.code === 'ERR_STREAM_PREMATURE_CLOSE') {
-          const result = {
-            ...collectionStream.getStats(),
-            aborted: true,
-          };
-          resolve(result);
-          return;
-        }
-        reject(err);
-        return;
-      }
+  try {
+    await pipeline(input, parseStream, docStream, collectionStream, abortSignal ? { signal: abortSignal } : undefined);
+  } catch (err: any) {
+    if (err.code === 'ERR_STREAM_PREMATURE_CLOSE') {
+      return {
+        ...collectionStream.getStats(),
+        aborted: true,
+      };
+    }
+    throw err;
+  }
 
-      resolve(collectionStream.getStats());
-    });
-  });
+  return collectionStream.getStats();
 }
```

Reply:
Thanks! It was indeed the optional abortSignal that was confusing it.